File: sdk-go/send_payloads_test.go
package primitive

import (
	"context"
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"

	primitiveapi "github.com/primitivedotdev/sdks/sdk-go/api"
)

type sendPayloadFixture struct {
	CanonicalInbound canonicalInbound `json:"canonical_inbound"`
	Send             []sendCase       `json:"send"`
	Reply            []replyCase      `json:"reply"`
	Forward          []forwardCase    `json:"forward"`
}

type canonicalInbound struct {
	ID             string            `json:"id"`
	EventID        string            `json:"event_id"`
	ReceivedAt     string            `json:"received_at"`
	Sender         canonicalAddress  `json:"sender"`
	ReplyTarget    canonicalAddress  `json:"reply_target"`
	ReceivedBy     string            `json:"received_by"`
	ReceivedByAll  []string          `json:"received_by_all"`
	Subject        *string           `json:"subject"`
	ReplySubject   string            `json:"reply_subject"`
	ForwardSubject string            `json:"forward_subject"`
	Text           *string           `json:"text"`
	Thread         canonicalInThread `json:"thread"`
	RawToHeader    string            `json:"raw_to_header"`
	RawDateHeader  string            `json:"raw_date_header"`
}

type canonicalAddress struct {
	Address string  `json:"address"`
	Name    *string `json:"name"`
}

type canonicalInThread struct {
	MessageID  *string  `json:"message_id"`
	InReplyTo  []string `json:"in_reply_to"`
	References []string `json:"references"`
}

type sendCase struct {
	Name                   string         `json:"name"`
	Input                  sendInputCase  `json:"input"`
	ExpectedBody           map[string]any `json:"expected_body"`
	ExpectedIdempotencyKey *string        `json:"expected_idempotency_key"`
}

type sendInputCase struct {
	From           string   `json:"from"`
	To             string   `json:"to"`
	Subject        string   `json:"subject"`
	BodyText       *string  `json:"body_text"`
	BodyHTML       *string  `json:"body_html"`
	InReplyTo      *string  `json:"in_reply_to"`
	References     []string `json:"references"`
	Wait           *bool    `json:"wait"`
	WaitTimeoutMs  *int     `json:"wait_timeout_ms"`
	IdempotencyKey *string  `json:"idempotency_key"`
}

type replyCase struct {
	Name                   string         `json:"name"`
	Input                  replyInputCase `json:"input"`
	ExpectedBody           map[string]any `json:"expected_body"`
	ExpectedIdempotencyKey *string        `json:"expected_idempotency_key"`
}

type replyInputCase struct {
	Text    string  `json:"text"`
	Subject *string `json:"subject"`
	From    *string `json:"from"`
}

type forwardCase struct {
	Name                     string           `json:"name"`
	Input                    forwardInputCase `json:"input"`
	ExpectedBodyMatch        map[string]any   `json:"expected_body_match"`
	ExpectedBodyTextContains []string         `json:"expected_body_text_contains"`
	ExpectedIdempotencyKey   *string          `json:"expected_idempotency_key"`
}

type forwardInputCase struct {
	To       string  `json:"to"`
	BodyText *string `json:"body_text"`
	Subject  *string `json:"subject"`
	From     *string `json:"from"`
}

type capturingSendAPI struct {
	request *primitiveapi.SendMailInput
	params  primitiveapi.SendEmailParams
}

func (s *capturingSendAPI) SendEmail(_ context.Context, request *primitiveapi.SendMailInput, params primitiveapi.SendEmailParams) (primitiveapi.SendEmailRes, error) {
	s.request = request
	s.params = params
	return &primitiveapi.SendEmailOK{
		Success: true,
		Data: primitiveapi.SendMailResult{
			ID:                   "sent-x",
			Status:               primitiveapi.SentEmailStatusSubmittedToAgent,
			QueueID:              primitiveapi.NilString{Null: true},
			Accepted:             []string{},
			Rejected:             []string{},
			ClientIdempotencyKey: "auto",
			RequestID:            "req",
			ContentHash:          "h",
		},
	}, nil
}

func loadSendPayloadFixture(t *testing.T) sendPayloadFixture {
	t.Helper()
	_, filename, _, _ := runtime.Caller(0)
	path := filepath.Join(
		filepath.Dir(filename), "..", "test-fixtures", "send-payloads", "cases.json",
	)
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	var f sendPayloadFixture
	if err := json.Unmarshal(data, &f); err != nil {
		t.Fatalf("decode fixture: %v", err)
	}
	return f
}

func deferenceString(p *string) string {
	if p == nil {
		return ""
	}
	return *p
}

func buildCanonicalReceivedEmail(c canonicalInbound) *ReceivedEmail {
	subject := deferenceString(c.Subject)
	messageID := deferenceString(c.Thread.MessageID)
	senderName := deferenceString(c.Sender.Name)
	replyTargetName := deferenceString(c.ReplyTarget.Name)
	rawDate := c.RawDateHeader

	senderHeader := c.Sender.Address
	if senderName != "" {
		senderHeader = senderName + " <" + c.Sender.Address + ">"
	}

	return &ReceivedEmail{
		ID:             c.ID,
		EventID:        c.EventID,
		ReceivedAt:     c.ReceivedAt,
		Sender:         ReceivedEmailAddress{Address: c.Sender.Address, Name: senderName},
		ReplyTarget:    ReceivedEmailAddress{Address: c.ReplyTarget.Address, Name: replyTargetName},
		ReceivedBy:     c.ReceivedBy,
		ReceivedByAll:  append([]string{}, c.ReceivedByAll...),
		Subject:        subject,
		ReplySubject:   c.ReplySubject,
		ForwardSubject: c.ForwardSubject,
		Text:           deferenceString(c.Text),
		Thread: ReceivedEmailThread{
			MessageID:  messageID,
			InReplyTo:  append([]string{}, c.Thread.InReplyTo...),
			References: append([]string{}, c.Thread.References...),
		},
		Raw: EmailReceivedEvent{
			ID: c.EventID,
			Email: Email{
				ID:         c.ID,
				ReceivedAt: c.ReceivedAt,
				Headers: EmailHeaders{
					From:      senderHeader,
					To:        c.RawToHeader,
					Subject:   c.Subject,
					Date:      &rawDate,
					MessageID: c.Thread.MessageID,
				},
			},
		},
	}
}

func captureRequestBody(t *testing.T, req *primitiveapi.SendMailInput) map[string]any {
	t.Helper()
	bytes, err := json.Marshal(req)
	if err != nil {
		t.Fatalf("marshal request: %v", err)
	}
	var out map[string]any
	if err := json.Unmarshal(bytes, &out); err != nil {
		t.Fatalf("unmarshal request: %v", err)
	}
	return out
}

func TestSharedSendPayloadFixtures(t *testing.T) {
	fixture := loadSendPayloadFixture(t)

	t.Run("send", func(t *testing.T) {
		for _, testCase := range fixture.Send {
			t.Run(testCase.Name, func(t *testing.T) {
				stub := &capturingSendAPI{}
				client := NewClientFromAPI(stub)

				params := SendParams{
					From:    testCase.Input.From,
					To:      testCase.Input.To,
					Subject: testCase.Input.Subject,
				}
				if testCase.Input.BodyText != nil {
					params.BodyText = *testCase.Input.BodyText
				}
				if testCase.Input.BodyHTML != nil {
					params.BodyHTML = *testCase.Input.BodyHTML
				}
				if testCase.Input.InReplyTo != nil || len(testCase.Input.References) > 0 {
					params.Thread = &SendThread{}
					if testCase.Input.InReplyTo != nil {
						params.Thread.InReplyTo = *testCase.Input.InReplyTo
					}
					if len(testCase.Input.References) > 0 {
						params.Thread.References = testCase.Input.References
					}
				}
				if testCase.Input.Wait != nil {
					params.Wait = testCase.Input.Wait
				}
				if testCase.Input.WaitTimeoutMs != nil {
					params.WaitTimeoutMs = *testCase.Input.WaitTimeoutMs
				}
				if testCase.Input.IdempotencyKey != nil {
					params.IdempotencyKey = *testCase.Input.IdempotencyKey
				}

				if _, err := client.Send(context.Background(), params); err != nil {
					t.Fatalf("Send returned error: %v", err)
				}

				body := captureRequestBody(t, stub.request)
				assertBodyEquals(t, body, testCase.ExpectedBody)
				assertIdempotencyKey(t, stub.params, testCase.ExpectedIdempotencyKey)
			})
		}
	})

	t.Run("reply", func(t *testing.T) {
		for _, testCase := range fixture.Reply {
			t.Run(testCase.Name, func(t *testing.T) {
				stub := &capturingSendAPI{}
				client := NewClientFromAPI(stub)
				email := buildCanonicalReceivedEmail(fixture.CanonicalInbound)

				rp := ReplyParams{BodyText: testCase.Input.Text}
				if testCase.Input.Subject != nil {
					rp.Subject = *testCase.Input.Subject
				}
				if testCase.Input.From != nil {
					rp.From = *testCase.Input.From
				}

				if _, err := client.Reply(context.Background(), email, rp); err != nil {
					t.Fatalf("Reply returned error: %v", err)
				}

				body := captureRequestBody(t, stub.request)
				assertBodyEquals(t, body, testCase.ExpectedBody)
				assertIdempotencyKey(t, stub.params, testCase.ExpectedIdempotencyKey)
			})
		}
	})

	t.Run("forward", func(t *testing.T) {
		for _, testCase := range fixture.Forward {
			t.Run(testCase.Name, func(t *testing.T) {
				stub := &capturingSendAPI{}
				client := NewClientFromAPI(stub)
				email := buildCanonicalReceivedEmail(fixture.CanonicalInbound)

				fp := ForwardParams{To: testCase.Input.To}
				if testCase.Input.BodyText != nil {
					fp.BodyText = *testCase.Input.BodyText
				}
				if testCase.Input.Subject != nil {
					fp.Subject = *testCase.Input.Subject
				}
				if testCase.Input.From != nil {
					fp.From = *testCase.Input.From
				}

				if _, err := client.Forward(context.Background(), email, fp); err != nil {
					t.Fatalf("Forward returned error: %v", err)
				}

				body := captureRequestBody(t, stub.request)
				for key, want := range testCase.ExpectedBodyMatch {
					got, ok := body[key]
					if !ok {
						t.Fatalf("body missing key %q", key)
					}
					if got != want {
						t.Fatalf("body[%q] = %v, want %v", key, got, want)
					}
				}
				bodyText, _ := body["body_text"].(string)
				for _, fragment := range testCase.ExpectedBodyTextContains {
					if !strings.Contains(bodyText, fragment) {
						t.Fatalf("body_text missing fragment %q\nfull body_text:\n%s", fragment, bodyText)
					}
				}
				assertIdempotencyKey(t, stub.params, testCase.ExpectedIdempotencyKey)
			})
		}
	})
}

func assertBodyEquals(t *testing.T, got, want map[string]any) {
	t.Helper()
	gotJSON, _ := json.Marshal(got)
	wantJSON, _ := json.Marshal(want)

	var gotNorm, wantNorm any
	_ = json.Unmarshal(gotJSON, &gotNorm)
	_ = json.Unmarshal(wantJSON, &wantNorm)

	if string(gotJSON) != string(wantJSON) {
		// JSON marshaling sorts keys alphabetically in Go; if string match
		// fails, use reflect-style comparison via re-marshaling as the
		// canonical form already, then compare.
		gotCanonical, _ := json.Marshal(gotNorm)
		wantCanonical, _ := json.Marshal(wantNorm)
		if string(gotCanonical) != string(wantCanonical) {
			t.Fatalf("body mismatch\n got: %s\nwant: %s", gotCanonical, wantCanonical)
		}
	}
}

func assertIdempotencyKey(t *testing.T, params primitiveapi.SendEmailParams, expected *string) {
	t.Helper()
	got, ok := params.IdempotencyKey.Get()
	if expected == nil {
		if ok {
			t.Fatalf("unexpected idempotency key %q", got)
		}
		return
	}
	if !ok {
		t.Fatalf("missing idempotency key, want %q", *expected)
	}
	if got != *expected {
		t.Fatalf("idempotency key %q, want %q", got, *expected)
	}
}

File: sdk-node/tests/api/send-payloads.test.ts
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { describe, expect, it, vi } from "vitest";
import {
  type EmailAnalysis,
  type EmailAuth,
  PrimitiveClient,
  type ReceivedEmail,
} from "../../src/index.js";

const __dirname = dirname(fileURLToPath(import.meta.url));
const FIXTURE_PATH = join(
  __dirname,
  "../../../test-fixtures/send-payloads/cases.json",
);

interface CanonicalInbound {
  id: string;
  event_id: string;
  received_at: string;
  sender: { address: string; name: string | null };
  reply_target: { address: string; name: string | null };
  received_by: string;
  received_by_all: string[];
  subject: string | null;
  reply_subject: string;
  forward_subject: string;
  text: string | null;
  thread: {
    message_id: string | null;
    in_reply_to: string[];
    references: string[];
  };
  raw_to_header: string;
  raw_date_header: string;
}

interface SendInputCase {
  from: string;
  to: string;
  subject: string;
  body_text?: string;
  body_html?: string;
  in_reply_to?: string;
  references?: string[];
  wait?: boolean;
  wait_timeout_ms?: number;
  idempotency_key?: string;
}

interface ReplyInputCase {
  text?: string;
  html?: string;
  from?: string;
  wait?: boolean;
}

interface ForwardInputCase {
  to: string;
  body_text?: string;
  subject?: string;
  from?: string;
}

interface SendCase {
  name: string;
  input: SendInputCase;
  expected_body: Record<string, unknown>;
  expected_idempotency_key: string | null;
}

interface ReplyCase {
  name: string;
  input: ReplyInputCase;
  expected_path: string;
  expected_body: Record<string, unknown>;
  expected_idempotency_key: string | null;
}

interface ForwardCase {
  name: string;
  input: ForwardInputCase;
  expected_body_match: Record<string, unknown>;
  expected_body_text_contains: string[];
  expected_idempotency_key: string | null;
}

interface FixtureFile {
  canonical_inbound: CanonicalInbound;
  send: SendCase[];
  reply: ReplyCase[];
  forward: ForwardCase[];
}

const FIXTURE = JSON.parse(readFileSync(FIXTURE_PATH, "utf8")) as FixtureFile;

const TEST_AUTH: EmailAuth = {
  spf: "pass",
  dmarc: "pass",
  dmarcPolicy: "reject",
  dmarcFromDomain: "example.com",
  dmarcSpfAligned: true,
  dmarcDkimAligned: true,
  dmarcSpfStrict: false,
  dmarcDkimStrict: false,
  dkimSignatures: [],
};

const TEST_ANALYSIS: EmailAnalysis = {};

function buildReceivedEmail(c: CanonicalInbound): ReceivedEmail {
  return {
    id: c.id,
    eventId: c.event_id,
    receivedAt: c.received_at,
    sender: c.sender,
    replyTarget: c.reply_target,
    receivedBy: c.received_by,
    receivedByAll: [...c.received_by_all],
    subject: c.subject,
    replySubject: c.reply_subject,
    forwardSubject: c.forward_subject,
    text: c.text,
    thread: {
      messageId: c.thread.message_id,
      inReplyTo: [...c.thread.in_reply_to],
      references: [...c.thread.references],
    },
    attachments: [],
    auth: TEST_AUTH,
    analysis: TEST_ANALYSIS,
    raw: {
      id: c.event_id,
      event: "email.received",
      version: "2025-12-14",
      delivery: {
        endpoint_id: "endpoint-1",
        attempt: 1,
        attempted_at: c.received_at,
      },
      email: {
        id: c.id,
        received_at: c.received_at,
        smtp: {
          helo: null,
          mail_from: "bounce@example.com",
          rcpt_to: [c.received_by],
        },
        headers: {
          message_id: c.thread.message_id,
          subject: c.subject,
          from: c.sender.name
            ? `${c.sender.name} <${c.sender.address}>`
            : c.sender.address,
          to: c.raw_to_header,
          date: c.raw_date_header,
        },
        content: {
          raw: {
            included: false,
            reason_code: "size_exceeded",
            max_inline_bytes: 0,
            size_bytes: 0,
            sha256: "0".repeat(64),
          },
          download: {
            url: "https://example.test/raw.eml",
            expires_at: "2026-01-01T01:00:00.000Z",
          },
        },
        parsed: {
          status: "complete",
          error: null,
          body_text: c.text,
          body_html: null,
          reply_to: [],
          cc: [],
          bcc: [],
          in_reply_to: [...c.thread.in_reply_to],
          references: [...c.thread.references],
          attachments: [],
          attachments_download_url: null,
        },
        analysis: TEST_ANALYSIS,
        auth: TEST_AUTH,
      },
    },
  };
}

const SUCCESS_RESPONSE = {
  success: true,
  data: {
    id: "sent-x",
    status: "submitted_to_agent",
    queue_id: null,
    accepted: [],
    rejected: [],
    client_idempotency_key: "auto",
    request_id: "req",
    content_hash: "h",
    idempotent_replay: false,
  },
};

function captureFetch(): {
  fetch: typeof fetch;
  captured: {
    path: string | null;
    body: Record<string, unknown> | null;
    idempotencyKey: string | null;
  };
} {
  const captured: {
    path: string | null;
    body: Record<string, unknown> | null;
    idempotencyKey: string | null;
  } = { path: null, body: null, idempotencyKey: null };
  const fetchMock = vi.fn<typeof fetch>(async (input) => {
    const request = input as Request;
    captured.path = new URL(request.url).pathname;
    captured.body = (await request.json()) as Record<string, unknown>;
    captured.idempotencyKey = request.headers.get("idempotency-key");
    return new Response(JSON.stringify(SUCCESS_RESPONSE), {
      status: 200,
      headers: { "content-type": "application/json" },
    });
  }) as typeof fetch;
  return { fetch: fetchMock, captured };
}

function buildClient(fetchImpl: typeof fetch): PrimitiveClient {
  return new PrimitiveClient({
    apiKey: "prim_test",
    baseUrl: "https://example.test/api/v1",
    fetch: fetchImpl,
  });
}

describe("shared send/reply/forward payloads", () => {
  for (const testCase of FIXTURE.send) {
    it(`send: ${testCase.name}`, async () => {
      const { fetch: fetchImpl, captured } = captureFetch();
      const client = buildClient(fetchImpl);

      const thread =
        testCase.input.in_reply_to !== undefined ||
        testCase.input.references !== undefined
          ? {
              ...(testCase.input.in_reply_to
                ? { inReplyTo: testCase.input.in_reply_to }
                : {}),
              ...(testCase.input.references
                ? { references: testCase.input.references }
                : {}),
            }
          : undefined;

      await client.send({
        from: testCase.input.from,
        to: testCase.input.to,
        subject: testCase.input.subject,
        ...(testCase.input.body_text !== undefined
          ? { bodyText: testCase.input.body_text }
          : {}),
        ...(testCase.input.body_html !== undefined
          ? { bodyHtml: testCase.input.body_html }
          : {}),
        ...(thread ? { thread } : {}),
        ...(testCase.input.wait !== undefined
          ? { wait: testCase.input.wait }
          : {}),
        ...(testCase.input.wait_timeout_ms !== undefined
          ? { waitTimeoutMs: testCase.input.wait_timeout_ms }
          : {}),
        ...(testCase.input.idempotency_key !== undefined
          ? { idempotencyKey: testCase.input.idempotency_key }
          : {}),
      });

      expect(captured.body).toEqual(testCase.expected_body);
      expect(captured.idempotencyKey).toBe(testCase.expected_idempotency_key);
    });
  }

  for (const testCase of FIXTURE.reply) {
    it(`reply: ${testCase.name}`, async () => {
      const { fetch: fetchImpl, captured } = captureFetch();
      const client = buildClient(fetchImpl);
      const email = buildReceivedEmail(FIXTURE.canonical_inbound);

      await client.reply(email, {
        ...(testCase.input.text !== undefined
          ? { text: testCase.input.text }
          : {}),
        ...(testCase.input.html !== undefined
          ? { html: testCase.input.html }
          : {}),
        ...(testCase.input.from !== undefined
          ? { from: testCase.input.from }
          : {}),
        ...(testCase.input.wait !== undefined
          ? { wait: testCase.input.wait }
          : {}),
      });

      // Path matters now: reply hits /emails/{id}/reply, not /send-mail.
      // A regression that wires reply back to send-mail would silently
      // build the wrong shape; the path assertion catches it before the
      // body assertion can give a misleading diff.
      expect(captured.path).toBe(`/api/v1${testCase.expected_path}`);
      expect(captured.body).toEqual(testCase.expected_body);
      expect(captured.idempotencyKey).toBe(testCase.expected_idempotency_key);
    });
  }

  for (const testCase of FIXTURE.forward) {
    it(`forward: ${testCase.name}`, async () => {
      const { fetch: fetchImpl, captured } = captureFetch();
      const client = buildClient(fetchImpl);
      const email = buildReceivedEmail(FIXTURE.canonical_inbound);

      await client.forward(email, {
        to: testCase.input.to,
        ...(testCase.input.body_text !== undefined
          ? { bodyText: testCase.input.body_text }
          : {}),
        ...(testCase.input.subject !== undefined
          ? { subject: testCase.input.subject }
          : {}),
        ...(testCase.input.from !== undefined
          ? { from: testCase.input.from }
          : {}),
      });

      expect(captured.body, testCase.name).toMatchObject(
        testCase.expected_body_match,
      );
      expect(typeof captured.body?.body_text, testCase.name).toBe("string");
      const bodyText = String(captured.body?.body_text ?? "");
      for (const fragment of testCase.expected_body_text_contains) {
        expect(bodyText, testCase.name).toContain(fragment);
      }
      expect(captured.idempotencyKey).toBe(testCase.expected_idempotency_key);
    });
  }
});
